### DISCUSSION TITLE
Optimizing ComfyUI with Parallelism and Scaling: A Proposal to Use Ray

### DISCUSSION THREAD

**Discussion by tempwswm** [Comfy-Org/ComfyUI#6647] [Ideas]
### Feature Idea

I’ve noticed that ComfyUI currently executes tasks sequentially, which limits its efficiency for various use cases. Introducing proper parallelism could lead to significant improvements:  

- **For personal use**, better parallelism would increase processing efficiency.  
- **For rented GPUs**, parallel execution can save costs by maximizing resource utilization.  
- **For large cluster service providers**, parallelism is essential to handle complex workloads effectively.  

To address this, I suggest considering **Ray** as the backbone for scaling and parallel execution in ComfyUI. Here are the reasons why Ray is a strong candidate:  

## Why Use Ray for Scaling in ComfyUI?  

1. **Ray is purpose-built for scaling AI and Python applications.**  
   It’s widely adopted in the AI community and has proven scalability features for both single-node and distributed applications.  

2. **Ray’s [objects](https://docs.ray.io/en/latest/ray-core/objects.html) system supports lazy evaluation and result caching.**  
   This can enhance execution efficiency for tasks that produce reusable outputs, such as certain node computations.  

3. **Similar paradigms: ComfyUI uses ‘nodes,’ and Ray uses ‘[actors](https://docs.ray.io/en/latest/ray-core/actors.html).’**  
   This similarity makes ComfyUI's workflow easier to adapt to Ray's model, thus simplifying integration. 

4. **Ray supports multiple [isolated](https://docs.ray.io/en/latest/ray-core/handling-dependencies.html#specifying-a-runtime-environment-per-task-or-per-actor) virtual environments.**  
   This feature is especially useful for resolving library/environment conflicts. Currently, ComfyUI manages different custom nodes that could rely on varying versions of the same library. While these do not necessarily cause runtime errors, they often lead to frustrating deprecation warnings. Ray’s isolation capabilities can mitigate this.  

5. **Community demand for better parallelism and distribution**  
   As mentioned earlier, the community needs improved parallelism. With Ray, it becomes feasible to distribute tasks across multiple GPUs, CPUs, or even clusters.  

### Existing Solutions

https://github.com/comfyanonymous/ComfyUI/issues/162
https://github.com/comfyanonymous/ComfyUI/issues/3538

## Key Resources / Links  
[ray github](https://github.com/ray-project/ray)
[ray core doc](https://docs.ray.io/en/latest/ray-core/walkthrough.html)
[SD example](https://docs.ray.io/en/latest/serve/tutorials/aws-neuron-core-inference-stable-diffusion.html)

By the way, it might also be worth considering **[UV](https://github.com/astral-sh/uv)** for installing custom nodes. UV is increasingly becoming the de facto standard for managing Python dependencies. Integrating UV could provide a cleaner and more efficient method for handling custom node installations.

**Comment by ltdrdata** [comment]
I think it would be best to discuss this in the discussion section.

**Comment by senlyu163** [comment]
Very interesting perspective! Does anyone have relevant experience?

**Reply by tempwswm** [reply]
I have, and am looking at the comfy source code.
But I'm not good enough to write pr anytime soon, that's why I'm proposing this, I need assistance😆

**Reply by cb88** [reply]
Xdit apparently also uses Ray... so if some major stuff is getting written it probably should also take into consideration adding advanced parallelism features xdit is developing.

**Comment by bot66** [comment]
Interesting, would you consider use `ray core` or `ray serve`

**Reply by lw309637554** [reply]
@tempwswm @bot66 I think we can refer to stable difussion on ray serve, the workload should not be too large https://docs.ray.io/en/latest/serve/tutorials/stable-diffusion.html

**Reply by lw309637554** [reply]
we can refer llama-factory  on ray
 https://github.com/hiyouga/LLaMA-Factory/pull/6542/files

**Reply by VikramxD** [reply]
would be very happy to help in this or anything that helps deploying and serving comfyUI better

**Reply by komikndr** [reply]
just use ray core

**Comment by komikndr** [comment]
I already trying to make custom nodes with ray.
There are numerous issue, mainly task barrier or synchronize, and ray not being called by the __main__. 
I am still figuring out how execution pipeline in comfy works

Task synchronization issue:
https://github.com/komikndr/raylight/blob/6130cad0c60618fb467748d1db7b85bfbb8864d6/src/raylight/nodes.py#L138

```python
        for actor in ray_actors:
            if parallel_dict["is_xdit"]:
                patched_futures.append(actor.patch_usp.remote())

            if parallel_dict["is_fsdp"]:
                patched_futures.append(actor.patch_fsdp.remote())

        ray.get(patched_futures)

        return (ray_actors,)
```

You see, `ray.get` is blocking and must wait untill every worker class, in this case `actor` in `ray_actors` to complete it's function. However Comfy suddenly jump into ksampler while the other worker hang waiting for the collective call. And when i am writing this i just realized i should make barrier before ksampler instead of in usp/fsdp patcher

**Reply by komikndr** [reply]
My Findings

1. **Model loading**  
   Loading a model into a Ray actor must be handled internally by the worker itself.  
   `load_diffusion_state_dict` should be placed inside the actor method.  
   If you simply copy from the **Load Diffusion** node into a Ray worker, Ray will serialize and deserialize the entire model object into the GCS (Global Control Service), which can spike RAM usage and cause OOM for large models.  

2. **Paths and libraries**  
   Paths and necessary libraries must be provided in `ray.init(runtime_env={modules})`.  
   This is because the libraries need to be transferred across ranks ,which is why there are multiple notes in my code explaining this.  

3. **SIGTERM warnings**  
   Since Ray is not being called from the main thread, it may raise warnings related to signal handling.  

4. **Torch distributed**  
   Be careful when assigning workers and communicating between them with Torch Distributed.  
   Ray isolates GPU workers from each other, so rank assignment must be carefully considered.  

5. **PORT usage**  
   Continue from point 4, ray will take quite a lot of ports, and also there are 2 address when using torch dist
   1 for torch dist comm, and the other for ray itself

6. **VRAM usage**  
   XDiT Ring parallel consumes much more VRAM compared to Ulysses.  

7. **Developer experience**  
   Using Ray is significantly more developer-friendly compared to `torch.multiprocessing`.  which is just unending pain  when it is not   being called by the `__main__`

Most of the issue stem from `__main__` and end user must be satisfied hence i cannot simply just ask them to `torch-run nproc -`

**Reply by komikndr** [reply]
The development is recoreded in my project README.md fully with VRAM usage and the it/s speed

https://github.com/komikndr/raylight

**Reply by komikndr** [reply]
Well it is not task sync issue, more a NCCL does not want to talk to each other

**Reply by komikndr** [reply]
There must also be a method, some kind of pre-load callback, where the model is initialized on the meta device for all ranks, and only rank 0 materializes the tensors and distributes them. If you want to apply something like FSDP, this is essential. Otherwise, if each rank gets its own model, memory usage will quickly balloon into OOM. Since at its core ray is its own Python interpreter. I have to swap the class method of model_patcher so it does not yet load the model into GPU

**Reply by komikndr** [reply]
Well, it is fixed, now you can use FSDP with LoRA without tanking memory. 

Basically, rank 0 loads as usual, but after `comfy.sd.load_diffusion_model()` I grab the `state_dict` and then convert `ModelPatcher.model` into the meta device. 

Broadcast every meta `ModelPatcher` into all ranks, then wrap and load the `state_dict` into FSDP. 
The trickiest part is LoRA, since we must convert it to `DTensor`, and hard set `lowvram_weight = True` in `ModelPatcher.load()`. 

The last problem to fully RayInit the ComfyUI model like clip, te, or vae must also be controlled by ray, so it can be registered into the remote cluster if you want to, as long as the ray cluster contains the folder for the model.

**Reply by komikndr** [reply]
if someone told me how on earth FSDP stuck in Comfy `class LoadedModel` that might be a good start

**Reply by komikndr** [reply]
## My Final Conclusion, Requirements for Converting to Ray

1. **Ray Worker**  
   - Ray is usually deployed on **Linux**.  
   - A **beta version** exists for Windows, but some Ray libraries still rely on **POSIX** instead of DOS standards, which limits full compatibility.

2. **DP Sampler (Data Parallel)**  
   - No need to modify the internal model’s forward logic, just assign each worker a **different seed and noise**.  
   - **Torch Distributed** is **not required**.

3. **SP Sampler (Sequence Parallel)**  
   - The transformer’s forward pass must be modified to support **sequence parallelism**, possibly through `transformer_options` to check if parallelism is enabled.  
     - If not enabled, simply skip the tensor sequence-splitting step.  
   - **Requires Torch Distributed**.  
   - GPU computation usually works in **powers of two**, so most sequences split evenly.  
   - Issues arise with **odd-numbered GPUs**, since you need to **pad odd tensors** and later **reduce** them back to the original latent size.  
     - My Qwen implementation suffers from this, some square images have mismatched sequence sizes between workers.

4. **Model Splitting Logic**  
   - **TP** is communication-heavy and splits dimensions differently than **FSDP**.  
     - FSDP treats sharding as `dim == 0` from Torch’s standpoint, allowing two GPUs to be both **sharded** and **sequence-parallel** at once.  
     - TP, however, requires **at least 4 GPUs**, two for TP dimensions and two for USP dimensions.  
   - **FSDP prefers uniform `dtype`**, but I’ve implemented a workaround to handle **non-uniform dtypes** (as found in scaled models).  
   - You’ll need to **manually drop** `current_model` inside the `model_management`.  
     - Check my repo for implementation , it works fine for now, but future compatibility is uncertain.

5. **Consumer Hardware**  
   - Parallelism love **high-speed communication**.  
     - The biggest bottleneck is **GPU P2P**.  
     - Without P2P, generation speed **drops significantly**.  
     - There’s a kernel patch available to enable P2P on consumer GPUs.  
     - NCCL not available on Windows
   - **PipeFusion** offers very low-communication-cost parallelism, but it requires major modifications to the **forward loop**.  
     - I believe this can be integrated via `comfy.patcher_extension.WrappersMP`, which I’m currently testing.

---

**Final Note**  
I designed my nodes to map **1:1** with **ComfyUI’s current library structure**.  
If this approach moves forward, I hope my project can be a helpful contribution.

**Reply by foreveremain** [reply]
Any update on this approach  ?

**Reply by komikndr** [reply]
SP and CFG parallel are already working fine. Nowadays i mostly fix FSDP Implementation, usually in this github repo https://github.com/komikndr/comfy-kitchen-distributed/

SP is mainly for DiT model that ran with uncond, like Wan + Lora.
While CFG is mainly for UNet

**Comment by komikndr** [comment]
## Native ComfyUI Ray implementation on `ray_worker` branch
https://github.com/komikndr/ComfyUI/tree/ray_worker

On testing, currently only USP is supported for WAN.

**Sequence Parallel (SP):**
Models without CFG (e.g., distilled WAN LoRA) can use sequence parallelism.

**No FSDP:**
FSDP on a dynamic system like ComfyUI is very unstable. It is usually only deployed in long-running, static pipelines (e.g., training or Diffusers). Raylight’s FSDP is experimental, and stacking it on top of USP (which is also experimental) increases instability. For now, I will let ComfyUI’s VRAM manager handle memory.

**Future work:**
- CFG parallelism
